rust/otap-dataflow/crates/engine/src/pipeline_ctrl.rs

```diff
@@ -107,35 +195,62 @@ impl PipelineCtrlMsgManager {
                         // No timers scheduled, wait indefinitely.
                         futures::future::pending::<()>().await;
                     }
-                }, if next_expiry.is_some() => {
-                    if let Some(Reverse((when, node_id))) = self.timers.pop() {
-                        // Only fire the timer if not canceled and not outdated.
-                        if let Some(timer_state) = self.timer_states.get_mut(&node_id) {
-                            if !timer_state.is_canceled && timer_state.scheduled_time == when {
-                                // Timer fires: handle expiration.
-                                if let Some(sender) = self.control_senders.get(&node_id) {
-                                    let _ = sender.send(NodeControlMsg::TimerTick {}).await;
-                                } else {
-                                    return Err(Error::InternalError {message: format!("No control sender for node: {node_id:?}")});
-                                }
-
-                                // Schedule next recurrence
-                                let next_when = Instant::now() + timer_state.duration;
-                                self.timers.push(Reverse((next_when, node_id)));
-                                timer_state.scheduled_time = next_when;
-                            } else if timer_state.is_canceled {
-                                // Clean up canceled timers
-                                let _ = self.timer_states.remove(&node_id);
-                            }
+                }, if next_earliest.is_some() => {
+                    let now = Instant::now();
+                    // Fire all due generic timers
+                    let senders = &self.control_senders;
+                    self.tick_timers.fire_due(now, |node_id| {
+                        if let Some(sender) = senders.get(node_id) {
+                            // best-effort; ignore send errors
+                            let _ = futures::executor::block_on(sender.send(NodeControlMsg::TimerTick {}));
                         }
-                    }
+                    });
+
+                    // Fire all due telemetry timers
+                    let senders = &self.control_senders;
+            self.telemetry_timers.fire_due(now, |node_id| {
+                        if let Some(sender) = senders.get(node_id) {
+                let _ = futures::executor::block_on(sender.send(NodeControlMsg::CollectTelemetry { metrics_reporter: self.metrics_reporter.clone() }));
```

Comment:
Could you confirm if it's safe to use `futures::executor::block_on` here? We have previously encountered issues with using `futures::executor::block_on` from within a current thread runtime. If the future that you're blocking on is not `Ready` when polled, then it hangs indefinitely.

For example, this code will hang indefinitely:

```rust
#[tokio::test] // Uses "current_thread" by default
async fn test_futures_executor_block_on() {
    let _ = futures::executor::block_on(tokio::time::sleep(Duration::from_millis(10)));
}
```
```

Reply:
Thanks for pointing out this problem. It is indeed a bad idea. I think we should perhaps find a way to detect this kind of issue in CI.